flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/CustomDistributionITCase.java

```diff
@@ -0,0 +1,230 @@
+import org.apache.flink.test.javaApiOperators.util.CollectionDataSets;
+import org.apache.flink.util.Collector;
+import org.junit.Test;
+
+
+import java.io.IOException;
+
+import static org.junit.Assert.fail;
+
+
+public class CustomDistributionITCase {
+	
+	@Test
+	public void testPartitionWithDistribution1() throws Exception{
+		/*
+		 * Test the record partitioned rightly with one field according to the customized data distribution
+		 */
+
+		ExecutionEnvironment env = ExecutionEnvironment.createLocalEnvironment();
+
+		DataSet<Tuple3<Integer, Long, String>> input1 = CollectionDataSets.get3TupleDataSet(env);
+		final TestDataDist1 dist = new TestDataDist1(1);
+
+		env.setParallelism(dist.getParallelism());
+
+		DataSet<Boolean> result = DataSetUtils.partitionByRange(input1, dist, 0).mapPartition(new RichMapPartitionFunction<Tuple3<Integer, Long, String>, Boolean>() {
+			@Override
+			public void mapPartition(Iterable<Tuple3<Integer, Long, String>> values, Collector<Boolean> out) throws Exception {
+				int partitionIndex = getRuntimeContext().getIndexOfThisSubtask();
+
+				for (Tuple3<Integer, Long, String> s : values) {
+					if ((s.f0 - 1) / 7 != partitionIndex) {
+						fail("Record was not correctly partitioned: " + s.toString());
+					}
+				}
+			}
+		});
+
+		result.output(new DiscardingOutputFormat()); 
+		env.execute();
+	}
+
+	@Test
+	public void testRangeWithDistribution2() throws Exception{
+		/*
+		 * Test the record partitioned rightly with two fields according to the customized data distribution
+		 */
+
+		ExecutionEnvironment env = ExecutionEnvironment.createLocalEnvironment();
+
+		DataSet<Tuple3<Integer, Long, String>> input1 = env.fromElements(
+						new Tuple3<>(1, 5L, "Hi"),
+						new Tuple3<>(1, 11L, "Hello"),
+						new Tuple3<>(2, 3L, "World"),
+						new Tuple3<>(2, 13L, "Hello World"),
+						new Tuple3<>(3, 8L, "Say"),
+						new Tuple3<>(4, 0L, "Why"),
+						new Tuple3<>(4, 2L, "Java"),
+						new Tuple3<>(4, 11L, "Say Hello"),
+						new Tuple3<>(5, 2L, "Hi Java"));
+
+		final TestDataDist2 dist = new TestDataDist2(2);
+
+		env.setParallelism(dist.getParallelism());
+
+		DataSet<Boolean> result = DataSetUtils.partitionByRange(input1.map(new MapFunction<Tuple3<Integer, Long, String>, Tuple3<Integer, Integer, String>>() {
+			@Override
+			public Tuple3<Integer, Integer, String> map(Tuple3<Integer, Long, String> value) throws Exception {
+				return new Tuple3<>(value.f0, value.f1.intValue(), value.f2);
+			}
+		}), dist, 0, 1).mapPartition(new RichMapPartitionFunction<Tuple3<Integer, Integer, String>, Boolean>() {
+			@Override
+			public void mapPartition(Iterable<Tuple3<Integer, Integer, String>> values, Collector<Boolean> out) throws Exception {
+				int partitionIndex = getRuntimeContext().getIndexOfThisSubtask();
+				for (Tuple3<Integer, Integer, String> s : values) {
+					
+					if ((s.f0 > partitionIndex + 1) || ((s.f0 == partitionIndex + 1) && (s.f1 > dist.rightBoundary[partitionIndex]))) {
```

Comment:
The test checks only one bucket boundary. We need to check 
- the upper boundary if `partitionIndex == 0`
- lower and upper boundaries if `0 < `partitionIndex` < 4`
- the lower boundary if `partitionIndex == 4.